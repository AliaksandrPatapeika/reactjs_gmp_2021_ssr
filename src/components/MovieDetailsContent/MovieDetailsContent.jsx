import {useRouter} from 'next/router';
import React, {useEffect} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import styled from 'styled-components';

import {getMovieById} from '../../actions/movies';
import {addDefaultSrc, delimiteredCost} from '../../utils';
import MovieDetailsLabel from '../MovieDetailsLabel';

const MovieDetailsContentContainer = styled.div`
  display: flex;
  padding: 30px 64px 64px 0;
`;
const MovieDetailsPoster = styled.img`
  align-self: flex-start;
  width: 294px;
  object-fit: cover;
`;
const MovieDetailsContentWrapper = styled.div`
  margin: 10px 0 0 63px;
  color: #ffffff;
  font-size: 18px;
  font-family: 'NetflixSans-Regular', sans-serif;
`;
const MovieDetailsTitleRating = styled.div`
  margin-bottom: 12px;
  font-size: 69px;
  font-family: 'TextaAlt-Regular', sans-serif;
`;
const MovieDetailsTitle = styled.span`
  margin-right: 20px;
`;
const MovieDetailsRating = styled.div`
  display: inline-block;
  width: 72px;
  height: 72px;
  border: 1.5px solid #ffffff;
  border-radius: 50%;
  color: #32CD32;
  font-size: 32px;
  text-align: center;
  vertical-align: middle;
  line-height: 66px;
`;
const MovieDetailsReleaseDateRuntime = styled.div`
  display: flex;
  margin: 18px 0;
  color: #f65261;
  font-size: 27px;
  & div:first-child {
    margin-right: 30px;
  }
`;
const Overview = styled.div`
  margin-top: 18px;
`;

const MovieDetailsContent = () => {
  const activeMovieDetailsMovie = useSelector((state) => state.movie.activeMovieDetailsMovie);
  const dispatch = useDispatch();
  const router = useRouter();
  const {id} = router.query;

  /* to display the movie details page after refresh */
  useEffect(() => {
    if (!activeMovieDetailsMovie && id) {
      dispatch(getMovieById(id));
    }
  }, [activeMovieDetailsMovie, dispatch, id]);

  return activeMovieDetailsMovie && (
    <MovieDetailsContentContainer>
      <MovieDetailsPoster
        src={activeMovieDetailsMovie.poster_path || '/img/noImage.png'}
        onError={addDefaultSrc}
        alt="movie details poster"
      />
      <MovieDetailsContentWrapper>
        <MovieDetailsTitleRating>
          <MovieDetailsTitle>{activeMovieDetailsMovie.title}</MovieDetailsTitle>
          <MovieDetailsRating title={`Vote count: ${activeMovieDetailsMovie.vote_count}`}>
            {activeMovieDetailsMovie.vote_average || '-'}
          </MovieDetailsRating>
        </MovieDetailsTitleRating>
        <div>{activeMovieDetailsMovie.tagline || 'No tagline'}</div>
        <MovieDetailsReleaseDateRuntime>
          <div title={`Release date: ${activeMovieDetailsMovie.release_date}`}>
            {new Date(
              activeMovieDetailsMovie.release_date
            ).getFullYear()}
          </div>
          <div>
            {activeMovieDetailsMovie.runtime || 0}
            {' '}
            min
          </div>
        </MovieDetailsReleaseDateRuntime>
        <MovieDetailsLabel label="Genres" value={activeMovieDetailsMovie.genres.join(', ')} />
        <MovieDetailsLabel label="Budget" value={delimiteredCost(activeMovieDetailsMovie.budget)} />
        <MovieDetailsLabel
          label="Revenue"
          value={delimiteredCost(activeMovieDetailsMovie.revenue)}
        />
        <Overview>{activeMovieDetailsMovie.overview || 'No overview'}</Overview>
      </MovieDetailsContentWrapper>
    </MovieDetailsContentContainer>
  );
};

export default MovieDetailsContent;
