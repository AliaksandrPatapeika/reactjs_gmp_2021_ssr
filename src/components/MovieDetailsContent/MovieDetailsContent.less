@import "../../less/variables";

.movieDetailsContentContainer {
	display: flex;
	padding: 30px 64px 64px 0;

	.movieDetailsPoster {
		align-self: flex-start;
		width: 294px;
		object-fit: cover;
	}

	.movieDetailsContent {
		margin: 10px 0 0 63px;
		color: @white;
		font-size: 18px;
		font-family: @NetflixSans-Regular;

		.movieDetailsTitleRating {
			margin-bottom: 12px;
			font-size: 69px;
			font-family: @TextaAlt-Regular;

			.movieDetailsTitle {
				margin-right: 20px;
			}

			.movieDetailsRating {
				display: inline-block;
				width: 72px;
				height: 72px;
				border: 1.5px solid @white;
				border-radius: 50%;
				color: @green;
				font-size: 32px;
				text-align: center;
				vertical-align: middle;
				line-height: 66px;
			}
		}

		.movieDetailsReleaseDateRuntime {
			display: flex;
			margin: 18px 0;
			color: @red;
			font-size: 27px;

			div:first-child {
				margin-right: 30px;
			}
		}

		.overview {
			margin-top: 18px;
		}
	}
}
