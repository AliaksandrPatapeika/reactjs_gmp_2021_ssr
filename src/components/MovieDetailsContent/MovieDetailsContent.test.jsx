import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import {configure, mount} from 'enzyme';
import {head} from 'lodash';
import React from 'react';
import {Provider, useSelector} from 'react-redux';
import renderer from 'react-test-renderer';
import configureMockStore from 'redux-mock-store';

import MockMovies from '../../tests/mocks/mockData';
import MovieDetailsContent from './MovieDetailsContent';

const mockedDispatch = jest.fn();

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: {id: '12'}
  })
}));

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: () => mockedDispatch
}));

configure({adapter: new Adapter()});

describe('MovieDetailsContent', () => {
  test('renders correctly', () => {
    const mockMovie = head(MockMovies);
    const mockStore = configureMockStore();
    const initialState = {
      movie: {
        activeMovieDetailsMovie: mockMovie
      }
    };
    const store = mockStore(initialState);

    useSelector.mockImplementation((callback) => callback(initialState));
    const movieDetailsContent = renderer
      .create(
        <Provider store={store}>
          <MovieDetailsContent title="title" />
        </Provider>
      )
      .toJSON();

    expect(movieDetailsContent).toMatchSnapshot();
  });

  test('loads data on init', () => {
    const initialState = {
      movie: {
        activeMovieDetailsMovie: null
      }
    };

    useSelector.mockImplementation((selectorFn) => selectorFn(initialState));
    mount(<MovieDetailsContent title="title" />);
    expect(mockedDispatch).toHaveBeenCalled();
  });
});
