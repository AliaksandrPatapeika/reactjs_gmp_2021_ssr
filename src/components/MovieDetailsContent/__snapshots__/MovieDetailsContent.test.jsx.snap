// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsContent renders correctly 1`] = `
<div
  className="MovieDetailsContent__MovieDetailsContentContainer-sc-1gunzxo-0 bwKGxc"
>
  <img
    alt="movie details poster"
    className="MovieDetailsContent__MovieDetailsPoster-sc-1gunzxo-1 bdIdZL"
    onError={[Function]}
    src="http://localhost:3000/img/01-pulp-fiction.jpg"
  />
  <div
    className="MovieDetailsContent__MovieDetailsContentWrapper-sc-1gunzxo-2 fhNbBK"
  >
    <div
      className="MovieDetailsContent__MovieDetailsTitleRating-sc-1gunzxo-3 ckiJyg"
    >
      <span
        className="MovieDetailsContent__MovieDetailsTitle-sc-1gunzxo-4 jKPqFV"
      >
        Pulp Fiction
      </span>
      <div
        className="MovieDetailsContent__MovieDetailsRating-sc-1gunzxo-5 iRaiVO"
        title="Vote count: 1195"
      >
        4.3
      </div>
    </div>
    <div>
      Mock tagline!
    </div>
    <div
      className="MovieDetailsContent__MovieDetailsReleaseDateRuntime-sc-1gunzxo-6 fjylqV"
    >
      <div
        title="Release date: 2004-03-15"
      >
        2004
      </div>
      <div>
        154
         
        min
      </div>
    </div>
    <div
      className="MovieDetailsLabel__MovieDetailsLabelWrapper-sc-35caid-0 wEuzg"
    >
      Genres
      : 
      <span
        className="MovieDetailsLabel__MovieDetailsValue-sc-35caid-1 cQRzbq"
      >
        Action & Adventure
      </span>
    </div>
    <div
      className="MovieDetailsLabel__MovieDetailsLabelWrapper-sc-35caid-0 wEuzg"
    >
      Budget
      : 
      <span
        className="MovieDetailsLabel__MovieDetailsValue-sc-35caid-1 cQRzbq"
      >
        55 000 000 $
      </span>
    </div>
    <div
      className="MovieDetailsLabel__MovieDetailsLabelWrapper-sc-35caid-0 wEuzg"
    >
      Revenue
      : 
      <span
        className="MovieDetailsLabel__MovieDetailsValue-sc-35caid-1 cQRzbq"
      >
        136 906 000 $
      </span>
    </div>
    <div
      className="MovieDetailsContent__Overview-sc-1gunzxo-7 hJWIDD"
    >
      Pulp Fiction is a 1994 American neo-noir black comedy crime film written and directed by Quentin Tarantino, who conceived it with Roger Avary.[4] Starring John Travolta, Samuel L. Jackson, Bruce Willis, Tim Roth, Ving Rhames, and Uma Thurman, it tells several stories of criminal Los Angeles. The title refers to the pulp magazines and hardboiled crime novels popular during the mid-20th century, known for their graphic violence and punchy dialogue.
    </div>
  </div>
</div>
`;
